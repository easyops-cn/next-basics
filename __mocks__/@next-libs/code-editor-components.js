import React from "react";

export function CodeEditorItem() {
  return <div>code editor</div>;
}
