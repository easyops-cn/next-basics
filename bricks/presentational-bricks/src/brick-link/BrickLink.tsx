import React from "react";
import { Tooltip } from "antd";
import { Link, LinkProps, GeneralIcon } from "@next-libs/basic-components";
import cssStyle from "./style.module.css";
import { isEmpty } from "lodash";
import { MenuIcon } from "@next-core/brick-types";
import classNames from "classnames";

export interface BrickLinkProps extends Pick<LinkProps, "replace" | "target"> {
  native?: boolean;
  label?: string;
  url?: string;
  href?: string;
  disabled?: boolean;
  tooltip?: string;
  tooltipProps?: Record<string, any>;
  handleClick?: () => void;
  notToJumpWhenEmpty?: boolean;
  icon?: MenuIcon;
  type?: "link" | "text";
  iconAlign?: "left" | "right";
  hideExternalIcon?: boolean;
  underLine?: boolean;
  labelColor?: string;
}

export function BrickLink(props: BrickLinkProps): React.ReactElement {
  const {
    native = false,
    label,
    url,
    href,
    tooltip,
    tooltipProps,
    type,
    iconAlign = "left",
    ...linkProps
  } = props;
  if (native) {
    // eslint-disable-next-line no-console
    console.warn(
      "presentational-bricks.brick-link `native` is deprecated, Please use `href` instead"
    );
  }

  const icon = props.icon ? (
    <GeneralIcon
      icon={props.icon}
      style={{
        ...(props.disabled && { color: "var(--color-disabled-text)" }),
        ...(label
          ? iconAlign === "left"
            ? { marginRight: 6 }
            : { marginLeft: 6 }
          : {}),
      }}
    />
  ) : null;

  let link;
  const labelEle = (
    <span
      className={props.underLine ? cssStyle.underLine : null}
      style={{
        borderBottomColor: props.labelColor,
        ...(props.labelColor ? { color: props.labelColor } : {}),
      }}
    >
      {label}
    </span>
  );
  const linkContent =
    iconAlign === "left" ? (
      <>
        {icon}
        {labelEle}
      </>
    ) : (
      <>
        {labelEle}
        {icon}
      </>
    );

  if (props.disabled) {
    link = <span className={cssStyle.disabledLink}>{linkContent}</span>;
  } else {
    const commonProps = {
      onClick: props.handleClick,
      className: classNames({
        [cssStyle.textLink]: type === "text",
      }),
    };
    if (props.notToJumpWhenEmpty && isEmpty(url)) {
      link = <a {...commonProps}>{linkContent}</a>;
    } else {
      link = native ? (
        <a href={url} {...linkProps} {...commonProps}>
          {linkContent}
        </a>
      ) : (
        <Link to={url} href={href} {...linkProps} {...commonProps}>
          {linkContent}
        </Link>
      );
    }

    if (props.target === "_blank" && !props.hideExternalIcon) {
      link = (
        <span>
          {link}
          <span className={cssStyle.externalIcon}>
            <GeneralIcon icon={{ lib: "fa", icon: "external-link-alt" }} />
          </span>
        </span>
      );
    }
  }

  return tooltip ? (
    <Tooltip {...tooltipProps} title={tooltip}>
      {link}
    </Tooltip>
  ) : (
    link
  );
}
