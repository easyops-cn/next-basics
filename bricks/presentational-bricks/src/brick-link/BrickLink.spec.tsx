import React from "react";
import { shallow } from "enzyme";
import { BrickLink, BrickLinkProps } from "./BrickLink";
import { GeneralIcon } from "@next-libs/basic-components";

jest.spyOn(console, "warn").mockImplementation(() => void 0);

describe("BrickLink", () => {
  it("should work", () => {
    const wrapper = shallow<BrickLinkProps>(
      <BrickLink
        native={true}
        label="aaa"
        url="/url"
        tooltip="tips.."
        icon={{
          lib: "antd",
          theme: "outlined",
          type: "file-search",
        }}
      />
    );

    expect(wrapper).toMatchSnapshot();
    wrapper.setProps({
      native: false,
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('target="_blank" should have external-link icon', () => {
    const wrapper = shallow<BrickLinkProps>(
      <BrickLink label="aaa" url="/url" target="_blank" />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should work when disabled", () => {
    const wrapper = shallow<BrickLinkProps>(
      <BrickLink
        label="aaa"
        disabled={true}
        icon={{
          lib: "antd",
          theme: "outlined",
          type: "file-search",
          color: "red",
        }}
      />
    );

    expect(wrapper.find("GeneralIcon").prop("style").color).toBe(
      "var(--color-disabled-text)"
    );
    expect(wrapper.find("Link").length).toBe(0);
    expect(wrapper.find("span").at(0).text()).toBe("<GeneralIcon />aaa");
  });

  it("should work when type is text", () => {
    const wrapper = shallow<BrickLinkProps>(
      <BrickLink label="aaa" url="/url" type="text" />
    );
    expect(wrapper.find(".textLink").length).toBe(1);
  });

  it("should work when iconAlign is right", () => {
    const wrapper = shallow<BrickLinkProps>(
      <BrickLink
        label="aaa"
        url="/url"
        type="text"
        iconAlign="right"
        icon={{
          lib: "antd",
          theme: "outlined",
          type: "file-search",
        }}
        target="_blank"
        underLine={true}
        labelColor={"red"}
      />
    );
    expect(wrapper.find("Link").childAt(1).props().icon).toEqual({
      lib: "antd",
      theme: "outlined",
      type: "file-search",
    });
    expect(wrapper.find(GeneralIcon).length).toBe(2);
    wrapper.setProps({
      hideExternalIcon: true,
    });
    expect(wrapper.find(GeneralIcon).length).toBe(1);
    expect(wrapper.find(".underLine").length).toBe(1);
    expect(wrapper.find(".underLine").prop("style")).toStrictEqual({
      borderBottomColor: "red",
      color: "red",
    });
  });
});
