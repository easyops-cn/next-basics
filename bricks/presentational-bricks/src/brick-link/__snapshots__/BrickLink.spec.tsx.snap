// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrickLink should work 1`] = `
<Tooltip
  arrowPointAtCenter={false}
  autoAdjustOverflow={true}
  mouseEnterDelay={0.1}
  mouseLeaveDelay={0.1}
  placement="top"
  title="tips.."
  transitionName="zoom-big-fast"
>
  <a
    className=""
    href="/url"
    icon={
      Object {
        "lib": "antd",
        "theme": "outlined",
        "type": "file-search",
      }
    }
  >
    <GeneralIcon
      icon={
        Object {
          "lib": "antd",
          "theme": "outlined",
          "type": "file-search",
        }
      }
      style={
        Object {
          "marginRight": 6,
        }
      }
    />
    <span
      className={null}
      style={
        Object {
          "borderBottomColor": undefined,
        }
      }
    >
      aaa
    </span>
  </a>
</Tooltip>
`;

exports[`BrickLink should work 2`] = `
<Tooltip
  arrowPointAtCenter={false}
  autoAdjustOverflow={true}
  mouseEnterDelay={0.1}
  mouseLeaveDelay={0.1}
  placement="top"
  title="tips.."
  transitionName="zoom-big-fast"
>
  <Link
    className=""
    icon={
      Object {
        "lib": "antd",
        "theme": "outlined",
        "type": "file-search",
      }
    }
    to="/url"
  >
    <GeneralIcon
      icon={
        Object {
          "lib": "antd",
          "theme": "outlined",
          "type": "file-search",
        }
      }
      style={
        Object {
          "marginRight": 6,
        }
      }
    />
    <span
      className={null}
      style={
        Object {
          "borderBottomColor": undefined,
        }
      }
    >
      aaa
    </span>
  </Link>
</Tooltip>
`;

exports[`BrickLink target="_blank" should have external-link icon 1`] = `
<span>
  <Link
    className=""
    target="_blank"
    to="/url"
  >
    <span
      className={null}
      style={
        Object {
          "borderBottomColor": undefined,
        }
      }
    >
      aaa
    </span>
  </Link>
  <span
    className="externalIcon"
  >
    <GeneralIcon
      icon={
        Object {
          "icon": "external-link-alt",
          "lib": "fa",
        }
      }
    />
  </span>
</span>
`;
